src/object/srv_ec_aggregate.c

```diff
@@ -560,31 +806,779 @@ agg_update_vos(struct ec_agg_entry *entry, bool write_parity)
 		}
 	}
 
-	ext_total = entry->ae_cur_stripe.as_suffix_ext ?
-		entry->ae_cur_stripe.as_extent_cnt - 1 :
-		entry->ae_cur_stripe.as_extent_cnt;
+	if (entry->ae_cur_stripe.as_ho_ext_cnt == 0 && agg_contained(entry)) {
+		epoch_range = ap->ap_epr;
+		recx.rx_idx = entry->ae_cur_stripe.as_stripenum *
+			obj_ec_stripe_rec_nr(entry->ae_oca);
+		recx.rx_nr = obj_ec_stripe_rec_nr(entry->ae_oca);
+		rc = vos_obj_array_remove(ap->ap_cont_handle,
+					  entry->ae_oid, &epoch_range,
+					  &entry->ae_dkey, &entry->ae_akey,
+					  &recx);
 
-	d_list_for_each_entry(agg_extent, &entry->ae_cur_stripe.as_dextents,
+	} else {
+		d_list_for_each_entry(ext, &entry->ae_cur_stripe.as_dextents,
+				      ae_link) {
+			int		erc = 0;
+			unsigned int	se = entry->ae_oca->u.ec.e_len *
+				entry->ae_oca->u.ec.e_k *
+				(entry->ae_cur_stripe.as_stripenum + 1);
+
+			if (ext->ae_orig_recx.rx_idx +
+			    ext->ae_orig_recx.rx_nr <= se) {
+				epoch_range.epr_lo = epoch_range.epr_hi =
+					ext->ae_epoch;
+
+				erc = vos_obj_array_remove(ap->ap_cont_handle,
+							   entry->ae_oid,
+							   &epoch_range,
+							   &entry->ae_dkey,
+							   &entry->ae_akey,
+							   &ext->ae_orig_recx);
+				if (erc)
+					D_ERROR("array_remov failed: "DF_RC"\n",
+					DP_RC(erc));
+				if (!rc && erc)
+					rc = erc;
+			}
+		}
+
+	}
+out:
+	return rc;
+}
+
+/* Retrieves the local replica extents from VOS, for the cells indicated
+ * by the bit_map.
+ */
+static int
+agg_fetch_local_extents(struct ec_agg_entry *entry, uint8_t *bit_map,
+			unsigned int cell_cnt, bool is_recalc)
+{
+	daos_iod_t		 iod = { 0 };
+	d_sg_list_t		 sgl = { 0 };
+	daos_recx_t		*recxs = NULL;
+	unsigned char		*buf = NULL;
+	struct ec_agg_param	*agg_param;
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	unsigned int		 i, j;
+	int			 rc = 0;
+
+	D_ALLOC_ARRAY(recxs, is_recalc ? cell_cnt : cell_cnt + 1);
+	if (recxs == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+
+	for (i = 0, j = 0; i < k; i++)
+		if (isset(bit_map, i)) {
+			recxs[j].rx_idx =
+			  entry->ae_cur_stripe.as_stripenum * k * len + i * len;
+			recxs[j++].rx_nr = len;
+		}
+	D_ASSERT(j == cell_cnt);
+
+	/* Parity is either updated (existing parity is updated),
+	 * or recalculated (generated from the entire stripe.
+	 *
+	 * Only need to fetch local parity if not recalculating it.
+	 */
+	if (!is_recalc) {
+		recxs[cell_cnt].rx_idx = PARITY_INDICATOR |
+				(entry->ae_cur_stripe.as_stripenum * len);
+		recxs[cell_cnt].rx_nr = len;
+	}
+
+	D_ALLOC_ARRAY(sgl.sg_iovs, cell_cnt + 1);
+	if (sgl.sg_iovs == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+
+	sgl.sg_nr =  is_recalc ? cell_cnt : cell_cnt + 1;
+	buf = (unsigned char *)entry->ae_sgl.sg_iovs[AGG_IOV_DATA].iov_buf;
+	for (i = 0; i < cell_cnt; i++) {
+		d_iov_set(&sgl.sg_iovs[i], &buf[i * len], len);
```

Comment:
The entry->ae_sgl.sg_iovs[AGG_IOV_DATA].iov_buf is k *len and cell_cnt is asserted to be <= k, since j <= k and D_ASSERT(j == cell_cnt).

for (i = 0, j = 0; i < k; i++)
                if (isset(bit_map, i)) {
                        recxs[j].rx_idx =
                          entry->ae_cur_stripe.as_stripenum * k * len + i * len;
                        recxs[j++].rx_nr = len;
                }
        D_ASSERT(j == cell_cnt);